### PR TITLE
feat: optimise where clause filtering

### LINKED ISSUE #2040: improvement: better disjoint where clause evaluation
Right now, when Electric knows about N shapes with where clauses, we're quite naive about figuring out which shapes does a given transaction fall into. For example, if we have N shapes over the same table with where clauses in the pattern of  `WHERE id = $1`.

We can improve the performance and reduce the CPU usage if we're smarter about it. For the example above, we can infer that if one of the where clauses is true, all the others are false and we don't need to evaluate them. Moreover, for `=` check, we can precompute a lookup structure to know in one lookup which where clauses of the set _can_ evaluate to `true` at all. If where clauses have `WHERE ID = $1 AND .... form`, then we can cut away definitely-false where clauses and execute only M << N checks instead.

The precomute can be done on a relatively rare event of shape creation (rare in relation to transaction processing).

### PR DESCRIPTION
**Pull request by robacourt**
Fixes #2040 and is estimated to improve Trigger.dev's replication stream processing by 400x

### CODE REVIEW

**Comment by balegas** [comment]
Exciting!

File: packages/sync-service/lib/electric/shapes/filter.ex [hunk 1]
```diff
@@ -50,51 +51,72 @@ defmodule Electric.Shapes.Filter do
     }
   end
 
-  def affected_shapes(%Filter{} = filter, %Relation{} = relation) do
+  def affected_shapes(%Filter{} = filter, change) do
+    shapes_affected_by_change(filter, change)
+  rescue
```
Comment (alco) [inline packages/sync-service/lib/electric/shapes/filter.ex:77]: Why is this needed? What kinds of errors do we expect to see here?

File: packages/sync-service/test/electric/shapes/filter_test.exs [hunk 2]
```diff
@@ -134,6 +134,59 @@ defmodule Electric.Shapes.FilterTest do
...
+        |> Enum.reduce(Filter.empty(), fn i, filter ->
+          Filter.add_shape(filter, i, Shape.new!("t1", where: "id = #{i}", inspector: @inspector))
+        end)
+
+      reductions =
+        reductions(fn ->
+          assert Filter.affected_shapes(filter, change("t1", %{"id" => "7"})) == MapSet.new([7])
+        end)
+
+      assert reductions < 500
```
Comment (alco) [inline packages/sync-service/test/electric/shapes/filter_test.exs:150]: Would be nice to see this specific number explain. What is the significance of it in this test? What should we do if this assertion fails?

**Review by alco — APPROVED** [review]
:muscle:

File: packages/sync-service/lib/electric/shapes/filter.ex (hunk 1 above)
Comment (robacourt) [reply packages/sync-service/lib/electric/shapes/filter.ex:77]: Super safety 😄 My reasoning is that if filtering crashes, all the shapes crash, so we need to be more careful than if this was shape specific code. There's a much more targeted `rescue` in `Table` that deals with record parsing errors which is much more likely to be hit. This one is more of a catch-all just to be extra safe. 

Types of error is could catch:
1. `Shape.is_affected_by_relation_change?` exceptions
2. change types the filter is not expecting are passed to `affected_shapes/2` 

These hopefully will never happen, but if they do, we'd rather the filter didn't crash.

File: packages/sync-service/test/electric/shapes/filter_test.exs (hunk 2 above)
Comment (robacourt) [reply packages/sync-service/test/electric/shapes/filter_test.exs:150]: Good point. Let me write a comment

**Review by icehaunter — APPROVED** [review]
Great job, exactly what I had in mind
